main.cc

```diff
@@ -446,6 +446,19 @@ class cql_config_updater {
     }
 };
 
+template <typename Func>
+inline auto defer_with_log_on_error(Func&& func) {
+    auto func_with_log = [func = std::forward<Func>(func)] () mutable {
+        try {
+            std::forward<Func>(func)();
+        } catch (...) {
+            startlog.error("Unexpected error on shutdown: {}", std::current_exception());
+            throw;
+        }
+    };
+    return deferred_action(std::move(func_with_log));
+}
+
```

Comment:
Why not simply always do that in `~deferred_action()`?
(or at least provide this function in defer.hh)